### Demonstration of four methods to sort a list: Identify the most efficient method

**Methods tested are as follows:**
1. `sorted()` function
2. Extended list slicing
3. `reverse()` method
4. `sort()` method

**1. Sort list using `sorted()` function.**

In [37]:
# Import package(s).
import timeit

# Create list.
lst_numbers = list(range(20))
print(lst_numbers)

# Sort list using sorted() function.
lst_numbers = sorted(lst_numbers, reverse=True)
print(lst_numbers)

# Create Python code for testing.
code = '''
lst_numbers = list(range(20))
lst_numbers = sorted(lst_numbers, reverse=True)
'''

# Execute Python code 1,000,000 times and return duration.
timeit.timeit(code, number=1000000)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
[19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]


2.0007217766371923

**2. Sort list using extended list slicing.**

In [32]:
# Create list.
lst_numbers = list(range(20))
print(lst_numbers)

# Sort list using extended list slicing.
lst_numbers = lst_numbers[::-1]
print(lst_numbers); print()

# Create Python code for testing
code = '''
lst_numbers = list(range(20))
lst_numbers = lst_numbers[::-1]
'''

# Execute Python code 1,000,000 times and return duration.
timeit.timeit(code, number=1000000)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
[19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]



0.8246288203396261

**3. Sort list using `reverse()` method.**

In [34]:
# Create list.
lst_numbers = list(range(20))
print(lst_numbers)

# Sort list using reverse() method.
lst_numbers.reverse()
print(lst_numbers)

# Create Python code for testing
code = '''
lst_numbers = list(range(20))
lst_numbers.reverse()
'''

# Execute Python code 1,000,000 times and return duration.
timeit.timeit(code, number=1000000)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
[19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]


0.662274113937201

**4. Sort list using `sort()` method.**

In [35]:
# Create list.
lst_numbers = list(range(20))
print(lst_numbers)

# Sort list using sort() method.
lst_numbers.sort(reverse=True)
print(lst_numbers)

# Create Python code for testing
code = '''
lst_numbers = list(range(20))
lst_numbers.sort(reverse=True)
'''

# Execute Python code 1,000,000 times and return duration.
timeit.timeit(code, number=1000000)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
[19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]


1.459661100782455

Based on the preceding results, the ranking of list sort methods by decreasing duration is as follows:
1. `reverse()` method: 0.662
2. Extended list slicing: 0.825
3. `sort()` method: 1.46
4. `sorted()` function: 2.001

Using the `reverse()` method is the most effective way to sort a list. The least effective way of sorting a list is the `sorted()` function. In fact, the `reverse()` method performs over **3x** faster than the `sorted()` function.